### DISCUSSION TITLE
Allow null conditional (?.) operator on left hand side of assignment

### DISCUSSION THREAD

**Discussion by mrjvdveen** [dotnet/csharplang#6072] [Language Ideas]
Consider this code:
```
if (Foo is { })
{
  Foo.Bar = "value";
}
```
I suggest introducing the possibility to write this like so:
`Foo?.Bar = "value";
`
Writing it like this, should do exactly the same as the first example. So if Foo is null, then the assignment never happens.

[Update(jcouv):] Detailed proposal write-up: https://github.com/dotnet/csharplang/issues/6045

**Comment by svick** [comment]
There has been a previous discussion about this at https://github.com/dotnet/csharplang/issues/1800, though that issue is now closed.

**Comment by Joe4evr** [comment]
And that issue was deemed to be a dupe of #34.

**Comment by svick** [comment]
@Joe4evr It was closed that way, but I'm not sure it was actually a duplicate. https://github.com/dotnet/csharplang/issues/34 is part of C# 8.0 and it doesn't let you do what's proposed here.

**Comment by Tom01098** [comment]
I dislike this because the whole line doesn't happen if the LHS is 'null', it is much clearer to wrap this in an `if` statement IMO.

**Comment by CyrusNajmabadi** [comment]
It makes sense to me, when you have `expr?.moreexpr`, the grammar and language is literally defiend where `.moreexpr` is simply a lower RHS node of the `?` (the `conditional access expression`).  So, if we allowed:  `expr?.name = whatever` then that would just be a node like so:

```
      ?
     / \
expr     =
        / \
   .name   whatever
```

So, like any conditional access expression, this would naturally fall out as "evaluate what is before the `?`.  if it `null`, do not evaluate what is after".

Feels fine and normal to me.  it's just something that didn't fall out with how we did the precedence for everything here.

**Comment by mrjvdveen** [comment]
I've had a brief discussion with @KathleenDollard at Techorama in the Netherlands two weeks ago about this feature. It is not a part of C# 8 right now and we both felt it deserved a feature request of it's own.

@Tom01098: I get your point. However I'm also allowed to write:
`foo?.Bar();`
This line will not execute if foo is null. In my view it would only be consistent to allow
`foo?.Prop = 'val';`

**Comment by YairHalberstadt** [comment]
@mrjvdveen
Can I suggest renaming this issue to "allow null conditional (?.) operator on left hand side of assignment"

**Comment by mrjvdveen** [comment]
> @mrjvdveen
> Can I suggest renaming this issue to "allow null conditional (?.) operator on left hand side of assignment"

Thanks for your input. I renamed this issue.

**Comment by DavidArno** [comment]
Having improved your title, would you mind removing the "*With the introduction of nullable reference types in C# 8 in mind*" line in your OP, please? The code, `if (Foo is { })` is a use of pattern matching and the rest of your post relates to the null conditional operator. None of it is related to NRTs.

**Comment by mrjvdveen** [comment]
> Having improved your title, would you mind removing the "With the introduction of nullable reference types in C# 8 in mind" line in your OP, please? The code, if (Foo is { }) is a use of pattern matching and the rest of your post relates to the null conditional operator. None of it is related to NRTs.

Thanks for your input. I have changed the OP. 
The first line of code is a way of doing a check for a non null object. It is relevant to the code and is correct by construction.

**Comment by alrz** [comment]
I think this has been brought up a couple times before, in short, it comes down to permitting `?.` as an lvalue
```cs
o?.Event += Handler;
```

**Comment by HaloFour** [comment]
#1106

**Comment by DavidArno** [comment]
@HaloFour, it's probably best to ignore #1106 though as it was completely derailed by the makeloft troll. Far better to have a new thread that restarts the discussion on the possible merits of allowing `?.` in lvalues.

**Comment by HaloFour** [comment]
@DavidArno 

There are other comments there as well.  We shouldn't let one bad actor wreck an entire thread.  Even if the conversation continues here I think it's good to link the two issues.

**Comment by mrjvdveen** [comment]
@HaloFour, I agree on having the link here. It documents the history of the discussion on this subject.

**Comment by jnm2** [comment]
I'd love to be able to write code like this:

```cs
(dataSource as BindingList)?.ListChanged -= OnListChanged;
dataSource = value;
(dataSource as BindingList)?.ListChanged += OnListChanged;
```

**Comment by HaloFour** [comment]
Would be nicely consistent with using `?.` to invoke mutator methods:

```cs
// works today
foo?.SetBar(CalculateBar());

// maybe?
foo?.Bar = CalculateBar();
```

**Comment by Leemyy** [comment]
I like this idea aswell, so I had a look through the Roslyn source to get a feeling for the ammout of work needed to implement this. 
From what I can tell, the syntax already accepts conditional access (`?.`) as an lvalue, so no changes will be required there. The Value checker is where it gets rejected, and where the error (`CS0131`) is produced. So some code will have to be added there to properly check and allow for `?.` as an lvalue. 
Of course that still won't magically make it work. Logic still needs to be added to the assignment lowering step, which will likely ammount to the majority of the work.

I've been looking to implement a compiler feature for a while now. I will have a go at implementing this and get back to you with my results. Even if this proposal gets rejected or my implementation proves inadequate, I can use the exercise.

**Comment by YairHalberstadt** [comment]
@leemyy

Have fun!

Be aware that PRs for language features that haven't been accepted are usually closed, so don't make a PR. Just leave it in your branch and add a link here.

If you need any help, check out the Roslyn gitter. (the amazing) @CyrusNajmabadi is usually happy to help , as am I, just less affectively.

Tagging @JaredPar, the Roslyn PM, to see if he thinks there's a chance a PR like this could get accepted.

**Comment by mrjvdveen** [comment]
Thanks for all the support.

@Leemyy : that's awesome. I'm curious to see how this pans out.
@YairHalberstadt: thanks for linking in people who could help out.

**Comment by Leemyy** [comment]
Hmm, while writing the Test cases I stubled upon an interesting corner case: Destructuring Assignments.
I was wondering whether the following should also be possible:
```c#
(left?.X, right?.X) = (1, 1);
```

**Comment by Leemyy** [comment]
I am thinking that it would be more consistent if it were possible, but maybe there are some aspects of it that I haven't considered?

**Comment by YairHalberstadt** [comment]
@leemyy

My advice is keep it simple for now, and forbid it. If this ever gets championed and bought up on a LDM, it will get discussed.

**Comment by canton7** [comment]
I think there's precedent in:

```cs
var (a, _) = (new object(), new object())
```

`new object()` is evaluated twice, even though one of the instances is discarded. I don't see `(left?.X, right?.X) = (new object(), new object())` as being any different if `left` or `right` is `null`.

But, as @YairHalberstadt says, leave this for the LDM.

**Comment by Leemyy** [comment]
You're right, this would add too much complexity to implement as part of the Prototype.

**Comment by Leemyy** [comment]
> 
> 
> I think there's precedent in:
> 
> ```cs
> var (a, _) = (new object(), new object())
> ```
> 
> `new object()` is evaluated twice, even though one of the instances is discarded. [...]

This raises another interesting question:
Should the simple conditional assignment evaluate the right side, if the left is null?

**Comment by mrjvdveen** [comment]
@Leemyy 
I'd not expect the right side to be evaluated. I'd keep it as close too a normal null check as possible.

**Comment by canton7** [comment]
> Should the simple conditional assignment evaluate the right side, if the left is null?

I'd argue no. `thing?.SetX(new object())` doesn't evaluate `new object()` if `thing` is `null`, so I think `thing?.X = new object()` should behave the same way.

**Comment by Leemyy** [comment]
I am thinking it should not, otherwise compound assignments would make no sense:
```c#
a?.b += x() * y;
```

Edit:
My thinking was that the decomposed form
```c#
if (a is object)
{
    a.b = a.b + (x() * y)
} 
else
{
    _ = a.b + (x() * y)
}
```
would try to access `a.b` even if `a` is null, but of course you could also decompose it as
```c#
if (a is object)
{
    a.b = a.b + (x() * y)
} 
else
{
    _ = x() * y
}
```
which would be fine.

**Comment by Leemyy** [comment]
I think if we agree that the RHS should not be evaluated when the LHS is null, that is another good case against allowing conditional assignment in destructuring assignments. It would make them way too hard to reason through, in my opinion.

**Comment by mrjvdveen** [comment]
@Leemyy : I completely agree with that

**Comment by Leemyy** [comment]
Just realized that we might also want to consider allowing conditional `++` and `--`. They are very similar to compound assignments and it could surprise users, that `a?.b += 1` is allowed but `a?.b++` and `++a?.b` are not.
I will leave this disallowed for now, but we should definitely consider it.

**Comment by ronnygunawan** [comment]
Another edge case:
```cs
a = (b?.c = d);
```

What would be the value of `b?.c = d` expression?

**Comment by mrjvdveen** [comment]
I'd say that d does not get evaluated. Therefor b.c is not assigned. This leads to a not being evaluated as well.

**Comment by ronnygunawan** [comment]
> I'd say that d does not get evaluated. Therefor b.c is not assigned. This leads to a not being evaluated as well.

`b?.c = d;` being not evaluated when `b` is null is ok, but when `a = ...` is not evaluated because the RHS contains skipped assignment, I think it would be confusing or unexpected.

Evaluating value of `(b?.c = d)` expression to the value of `d` would be a more predictable behavior.

**Comment by CyrusNajmabadi** [comment]
Because it may not evaluate, it's result would be a nullable of whatever value it might produce.  

i.e. if i do this:

```
a = b?.SetC(d);
```

Then if `b` is null, then `a` is assigned the null value of the type of the expr.  SO i would expect the same for:

```
a = (b?.C = d)
```

If `b` is null, then there is no assignment and hte entire expression on the rhs is effectively null which is then assigned into 'a'.

**Comment by Leemyy** [comment]
It looks like there are three ways to interpret `a = b.C?.D = x.Y()`:
(In all these `RetY` is assumed to be the return type of `x.Y()`)

---
A)
```c#
if (b.C is {} b_C)
{
    RetY temp = x.Y();
    b_C.D = temp;
    a = temp;
}
```
---
B)
```c#
RetY temp = x.Y();
if (b.C is {} b_C)
{
    b_C.D = temp;
}
a = temp;
```
---
C)
```c#
RetY? temp = null;
if (b.C is {} b_C)
{
    temp = x.Y()
    b_C.D = temp;
}
a = temp;
```
---
I don't like approach A, because it may not assign to `a` at all. This seems very counter-intuitive to me, as just having `a = b.C?.D` *is* guaranteed to assign to `a`. I am also not sure how this approach would scale if we had an even longer assignment, like `a = b.C?.D = e?.F = x.Y()`. Would `b.C` be checked for `null` first, or would `e` be checked first? Note that this makes a difference, because `C` might have side effects (though it arguably shouldn't).

Approach B is a bit more well behaved, but it only requres `a` to be of type `RetY`, where just having `a = b.C?.D` would require it to be of type `RetY?` instead. It's also confusing because it always evaluates `x.Y()`, while `b.C?.D = x.Y()` would only evaluate it if `b.C` is not `null`. 
Beyond that, it is also unclear what `a?.Z = b.C?.D = x.Y()` should do. Should it still evaluate `x.Y()`, even if both `a` and `b.C` are null? Or should it only evaluate it, if there is at least one receiver that is not `null`?

This brings me to approach C, which I like the best. Its individual parts resolve to the same types as they would if you just did `b.C?.D = x.Y()` followed by `a = b.C?.D`, and it has the same effect, too (except for the double evaluation of `b.C?.D`). It is also really straightforward how to extend this pattern to even longer chains, like `a = b.C?.D = e?.F = x.Y()`.
One thing about long chained assignments like `a = b.C?.D = e?.F = x.Y()` that is still unclear is whether the `null`-check on `b.C`, or on `e` should happen first. Checking `b.C` first allows us to skip both checking `e` and assigning to `e.F`. As I have already explained above, however, I don't think skipping any of the assignments is a good idea. As such, I think evaluating the the assignments from right to left is the way to go.
Doing things this way also comes with another bonus: You can know what any of the individual assignments will do, without having to consider anything but the immedeate left, and right.

TL;DR: I agree with @CyrusNajmabadi

**Comment by Leemyy** [comment]
> [...] I think evaluating the the assignments from right to left is the way to go. [...]

I just realized that this contradicts our rule that the RHS of a conditional assignment should not get evaluated, if the LHS is `null`. I do, however, still think that chained assignments should not skip any assignments, unless they are explicitly conditional. Otherwise the following snippet would not be guaranteed to assign `s`:
```c#
string? s;
a?.B = s = "uncertainty"
```
Changing the order of the assignments, however, *does* guarantee that `s` is assigned:
```c#
string? s;
s = a?.B = "uncertainty"
```
This order dependence is likely to confuse loads of people; another reason why I want to avoid it.

**Comment by jaredpar** [comment]
> Tagging @jaredpar, the Roslyn PM, to see if he thinks there's a chance a PR like this could get accepted.

Technically still a Dev but ... 

A PR to Roslyn directly would not be accepted. This is a language change and hence needs to go through LDM approval first. They essentially declare what is / isn't legal in the language and the compiler implements the legal portions.

The next step would be a formal proposal for the feature. The description of allow `?.` is not sufficient. It needs to deeper in terms of describing the behavior. For instance the spec should be able to articulate why 

```csharp
a?.b = 42; // legal
M(ref a?.b); // illegal
```
If for instance `a?.b` was described as an lvalue / location then I would expect `ref` to be used on it as well. 

Further interesting questions:

- For an expression of form `a?.b = c` what is the resulting value? After all assignment is an expression.
- For an expression of the `a?.b = c` what sets of expressions are legal for `a`? Are method calls legal for example (think about `ref` returns). What about `struct` returns?

**Comment by mrjvdveen** [comment]
@jaredpar Thanks for clarifying that. So how would the formal proposal be made? Is this something that can be done by the opensource community? And if so, what would the proposal look like?

I also think that the discussion that has been going on so far can be a great help towards writing a formal proposal.

**Comment by Leemyy** [comment]
I am curious about this process aswell.
I did find explanations of [the design process](https://github.com/dotnet/csharplang#design-process), aswell as [the proposal lifecycle](https://github.com/dotnet/csharplang/tree/master/proposals#c-language-proposals), and [what a proposal should look like](https://github.com/dotnet/csharplang/blob/master/proposals/proposal-template.md).

From the looks of it, only members of the language design team can bring a feature request into one of the language design meetings (aka. champion it). I am not sure on what grounds feature requests are championed, though. I am also uncertain whether it makes sense to make a draft spec before an issue gets championed, or whether the speccing is something that is done by the proposal champion, after it was approved in an LDM. I think that it would make sense to at least have some kind of summary of the different options we have already discussed, which might aswell be in the form of a draft spec.

**Comment by jaredpar** [comment]
> Is this something that can be done by the opensource community?

Absolutely. Generally the proposal is done either in an issue or by submitting a proposal via PR. If you look at the history of the PRs in this repo you will see a number of examples. 

> And if so, what would the proposal look like?

Look under the proposals folder in this repo. There are a good number there which have the general structure. There are also some well written ones under the csharp 8 and 7.* folders. 

> . I am not sure on what grounds feature requests are championed, though.

Mostly just getting a member of LDM interested in the feature. Generally if there is a decent proposal it will get championed.

**Comment by michael-hawker** [comment]
We [do this all the time in the toolkit for templated controls](https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/068faf491fcb040ef7fafc15ce18710946eadd57/Microsoft.Toolkit.Uwp.UI.Controls/TabView/TabView.cs#L92-L115), where we need to register event handlers:

```cs
            if (_tabItemsPresenter != null)
            {
                _tabItemsPresenter.SizeChanged -= TabView_SizeChanged;
            }

            if (_tabScroller != null)
            {
                _tabScroller.Loaded -= ScrollViewer_Loaded;
            }

            _tabContentPresenter = GetTemplateChild(TabContentPresenterName) as ContentPresenter;
            _tabViewContainer = GetTemplateChild(TabViewContainerName) as Grid;
            _tabItemsPresenter = GetTemplateChild(TabsItemsPresenterName) as ItemsPresenter;
            _tabScroller = GetTemplateChild(TabsScrollViewerName) as ScrollViewer;

            if (_tabItemsPresenter != null)
            {
                _tabItemsPresenter.SizeChanged += TabView_SizeChanged;
            }

            if (_tabScroller != null)
            {
                _tabScroller.Loaded += ScrollViewer_Loaded;
            }
```

It'd be nice to condense this down to this:

```cs
            _tabItemsPresenter?.SizeChanged -= TabView_SizeChanged;
            _tabScroller?.Loaded -= ScrollViewer_Loaded;

            _tabContentPresenter = GetTemplateChild(TabContentPresenterName) as ContentPresenter;
            _tabViewContainer = GetTemplateChild(TabViewContainerName) as Grid;
            _tabItemsPresenter = GetTemplateChild(TabsItemsPresenterName) as ItemsPresenter;
            _tabScroller = GetTemplateChild(TabsScrollViewerName) as ScrollViewer;

            _tabItemsPresenter?.SizeChanged += TabView_SizeChanged;
            _tabScroller?.Loaded += ScrollViewer_Loaded;
```

I also see this tied with the property assignment as well, I forget my exact use-case the other day, but wanted to do something similar where I may try and grab a reference, and if I do set the property:

```cs
    var getobj = LookupThing();
    getobj?.Property = "Some Value";
```

**Comment by jnyrup** [comment]
Would this also handle null conditional assignment on set-only properties?

```cs
using System;

public class MyClass
{
    private string property;
    public string Property
    { 
        set => property = value;
    }
    
    private void MyMethod()
    {
        // use property
    }
}    

public static class Program 
{
    public static void Main() 
    {
        object obj = new MyClass();
        
        //CS0154: The property or indexer 'MyClass.Property' cannot be used in this context because it lacks the get accessor

        (obj as MyClass)?.Property = ""; 
    }
}
```
https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOwD4AEBMAjALABQZuAzAAQHUCyAngMIA20UZA3mdX9QAcEASwBuEeLUIAGQQgD2AxDEYBuXvypTZABQVKEKjXy7Vj/PlDgxqAXgB8cxcsZ3q41ujjrSFgL7m5kJiEnC0ACwMjPTWABbyACYAFACU5jy+FvwA9NnU6FZOBkaZfAGkfhZkFPg03OYE+Gal1BlZ/PLAAFZwAMY2nV1umHAA7lFsHKk+7XzmFkmD1NAT7FBQKQD8AHR6zoauttQARMcz/OV+QA===

**Comment by Rekkonnect** [comment]
> Would this also handle null conditional assignment on set-only properties?

The error in that code sample is only produced because, currently, the compiler expects that with `?.` you're only trying to get a value, not assign.

Properties with setters should allow `?.` assignment, regardless of whether they contain a getter or not.

**Comment by CyrusNajmabadi** [comment]
I'll champion this as well.

**Comment by Rekkonnect** [comment]
Does this proposal also handle the case of using the `?[` operator for indexing arrays, or other objects with custom indexers in the LHS?

**Comment by Xyncgas** [comment]
Is it escaping type safety? something must've gone wrong if you are looking for members in a variable that is null as a consumer. but if you are providing the value you can do this in generic and write once apply to all so it seems to me that it's escaping dealing with the consequences of null in consumer's side

it's like shorthand for ```try {foo.bar = 5} catch { }``` or
```
   #IgnoreTypeSafety
   foo.bar = 5
   #endIgnoreTypeSafety
```

regardless I still want this feature to be implemented because I want to be lazy and write unsafe code, what can possibly go wrong, at the end of the day I will be blamed for bugs anyways

**Comment by Rekkonnect** [comment]
It is not a shorthand for either of those cases. The null-conditional operator for any member access operator (`.`, `[`) *should* behave like the following:

```csharp
if (foo is not null)
   foo.bar = 5;
```

No exceptions involved, because they're too expensive. No type unsafety either, you always account for `null` when dealing with references.

> something must've gone wrong if you are looking for members in a variable that is null as a consumer

Even with your nullability settings and your strong usage of the `!` operator, you **cannot** (at least reliably) get rid of null checks. Whether something's gone wrong or not, it's not the compiler's duty to cleverly eliminate null checks. If this ever happens, it will be during JIT compilation, and I don't even know if it does so, let alone how frequently and reliably.

**Comment by Xyncgas** [comment]
Maybe we can get rid of null in c# by opt-in feature, instead of giving nullable to everything, and having variables points to nullptr they should have a pointer to a static object that's the default state. I feel annoyed when I forgot to check for null and things break during testing too, only if things were never null I can be happier, I understand I can also take a walk outside and stop returning null in new codes I write it isn't a big deal for me

**Comment by Rekkonnect** [comment]
This is just not happening. `null` will be part of the language since it's been since day one, and it's also useful in certain scenarios.

You can use the nullability tooling available since C# 8.0/.NET Core 3.1(?) and respect it like you personally desire. Other than that, if you _know_ your values aren't `null`, you shouldn't care. That's a general development tip, to always properly track what values you're passing. `null` values aren't the only problematic ones.

**Comment by Xyncgas** [comment]
New rules ok
imagine you have the object be ```default``` instead of ```null```, which I think microsoft made the decision too soon to say ```default``` is ```null``` for reference types, because it's gonna break when we change this behavior later and they saw this coming too people have been saying let's get rid of null and RUST did, but anyways I think it would be ideal to let ```default``` to actually be type ```T``` that lives as an static object in runtime and anytime you try to do something like ``` ObjectA.b ``` it's just not possible to be ```null``` because ```ObjectA``` by default has a reference to the default state of its type if you've never set it or forgot to

**Comment by Rekkonnect** [comment]
Such a breaking change is probably never going to happen. I like the idea of having a "default" static object, which from what I assume would have to be constructed by the default instance constructor that will be introduced for that feature. It will have to be triggered in some other way though, like `T.default`, since `default(T)` is already valid syntax, despite being simplified to `default`.

This discussion however should be made in the respective proposal.

**Comment by Haukinger** [comment]
I'd always prefer seeing that something's going wrong instead of having it going havoc silently. That's what `null` and `NullReferenceException` are for - notifying you that something gone awry.

If I have code that accepts optional parameters or values, that code has to check for `null` and react accordingly.
Also, if I have code that doen't accept optional parameters or values, that code, too, has to check for `null` just to be on the safe side and then react accordingly.

Also, `if (value == null)` and much more so `value?.` is a lot more concise than verbose alternatives like `Nullable` (i.e. `if (value.HasValue) ... value.Value`). Conciseness is all what `?.` is about, mind you.

**Comment by Xyncgas** [comment]
@jnyrup 

In your case I would expect it not work (by nothing happens) since the object isn't null but it's the property that wasn't able to be set

**Comment by Xyncgas** [comment]
Any Updates?

**Comment by CyrusNajmabadi** [comment]
@Xyncgas I've championed this.

**Comment by legistek** [comment]
I would love to see this!

**Comment by 333fred** [comment]
I'm converting this issue to a discussion, as https://github.com/dotnet/csharplang/issues/6045 has the template and proposal.
